compiler-core/src/parse.rs

```diff
@@ -1395,6 +1395,21 @@ where
         }
     }
 
+    // checks if we have a function call after we find an invalid left paren
+    // Warning: this function will consume the tokens
+    // Returns the location of the right paren if we have a function call
+    fn is_invalid_call_in_case_clause_guard(&mut self) -> Option<u32> {
+        if self.parse_fn_args().is_err() {
```

Comment:
As the comment says this could consume some characters but then fail, at which point the parser is left in an "ok" state at the level up. This could result in some invalid code being "successfully" parsed.

The parser must always remain in an error state once it encounters an error.

How about making this function return a parse error instead? It would ensure it is always correct, and it removes some of the duplicated logic from the sites where this function is called